src/main/scala/rules/Executor.scala

```diff
@@ -63,74 +64,136 @@ object executor extends ExecutionRules {
+          eval(s1, ce.condition, IfFailed(ce.condition), v)((s2, tCond, v1) =>
+            /* Using branch(...) here ensures that the edge condition is recorded
+             * as a branch condition on the pathcondition stack.
+             */
+            brancher.branch(s2, tCond, v1)(
+              (s3, v3) =>
+                exec(s3, ce.target, ce.kind, v3, joinPoint)((s4, v4) => {
+                  SymbExLogger.currentLog().closeScope(sepIdentifier)
+                  Q(s4, v4)
+                }),
+              (_, _)  => {
                 SymbExLogger.currentLog().closeScope(sepIdentifier)
-                Q(s4, v4)
-              }),
-            (_, _)  => {
-              SymbExLogger.currentLog().closeScope(sepIdentifier)
-              Success()
-            }))
-
-      case ue: cfg.UnconditionalEdge[ast.Stmt, ast.Exp] =>
-        val s1 = handleOutEdge(s, edge, v)
-        exec(s1, ue.target, ue.kind, v)(Q)
+                Success()
+              }))
+
+        case ue: cfg.UnconditionalEdge[ast.Stmt, ast.Exp] =>
+          val s1 = handleOutEdge(s, edge, v)
+          exec(s1, ue.target, ue.kind, v, joinPoint)(Q)
+      }
     }
   }
 
   private def follows(s: State,
                       edges: Seq[SilverEdge],
                       @unused pvef: ast.Exp => PartialVerificationError,
-                      v: Verifier)
+                      v: Verifier,
+                      joinPoint: Option[SilverBlock])
                      (Q: (State, Verifier) => VerificationResult)
                      : VerificationResult = {
 
     if (edges.isEmpty) {
       Q(s, v)
     } else if (edges.length == 1) {
-      follow(s, edges.head, v)(Q)
-    } else {
-      val uidBranchPoint = SymbExLogger.currentLog().insertBranchPoint(edges.length)
-      val res = edges.zipWithIndex.foldLeft(Success(): VerificationResult) {
-        case (fatalResult: FatalResult, _) => fatalResult
-        case (_, (edge, edgeIndex)) =>
-          if (edgeIndex != 0) {
-            SymbExLogger.currentLog().switchToNextBranch(uidBranchPoint)
+      follow(s, edges.head, v, joinPoint)(Q)
+    } else if (edges.length == 2) {
+      val edge1 = edges.head.asInstanceOf[ConditionalEdge[ast.Stmt, ast.Exp]]
+      val edge2 = edges.last.asInstanceOf[ConditionalEdge[ast.Stmt, ast.Exp]]
+
+      assert(edge1.source == edge2.source)
+      val branchPoint = edge1.source
+
+      s.methodCfg.joinPoints.get(branchPoint) match {
+        // If a join point was found, the config argument is enabled,
+        // and the source of both conditional edges was a statement block
+        // (we are at an if-statement), we may join the resulting branches again.
+        case Some(newJoinPoint) if
+          Verifier.config.moreJoins() &&
+          edge1.isInstanceOf[ConditionalEdge[ast.Stmt, ast.Exp]] &&
+          edge2.isInstanceOf[ConditionalEdge[ast.Stmt, ast.Exp]] &&
+          edge1.source.isInstanceOf[StatementBlock[ast.Stmt, ast.Exp]] &&
```

Comment:
Why are the restrictions on lines 127 and 128 necessary? Please explain in code comments.